atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java

```diff
@@ -244,13 +244,17 @@ private void testDataSource(HikariDataSource dataSourcePool) throws SQLException
      * @throws SQLException if connection is invalid and cannot be used.
      */
     private void testConnection(Connection connection) throws SQLException {
-        try (Statement stmt = connection.createStatement();
-                ResultSet rs = stmt.executeQuery(connConfig.getTestQuery())) {
-            if (!rs.next()) {
-                throw new SQLException(String.format(
-                        "Connection %s could not be validated as it did not return any results for test query %s",
-                        connection, connConfig.getTestQuery()));
+        if (connConfig.getTestQuery() != null) {
+            try (Statement stmt = connection.createStatement();
+                    ResultSet rs = stmt.executeQuery(connConfig.getTestQuery())) {
+                if (!rs.next()) {
+                    throw new SQLException(String.format(
+                            "Connection %s could not be validated as it did not return any results for test query %s",
+                            connection, connConfig.getTestQuery()));
+                }
             }
+        } else if (!connection.isValid(connConfig.getSocketTimeoutSeconds())) {
+            throw new SQLException(String.format("Connection %s is not valid", connection));
```

Comment:
Will let this through to the keeper...

Reply:
TBH it is unclear to me why we are doing this at all. We get no guarantees the connection is actually still good by the time we actually us it, and this adds a huge overhead to retrieving a connection from the pool. Profiling an operational stack this check is one of the most expensive queries in total CPU - even though it does nothing.

Reply:
Yes, the `SELECT 1 FROM DUAL` / `SELECT 1` are expensive, mostly in the additional round trips. The JDBC 4 `isValid()` is supposed to make this much cheaper, but as mentioned with earlier versions of some JDBC drivers there were issues with connection corruption. There's probably also a missing optimization  here to set the row fetch size to 1 for the test query prepared statement to avoid constant reallocation of the driver internal array buffers used for query request and responses.

I don't recall the full history of why this additional set of connection validations was added on top of Hikari, but I think #2301 and #2142 cover some of the Chesterton's fence archeology. Those too seem to point at issues with some of the ways where interruptions are used as part of concurrent queries either due to hitting an application enforced timeout (e.g. search or query timeout limit), and cross thread communications.

[JDBC and Multithreading](https://docs.oracle.com/en/database/oracle/oracle-database/21/jjdbc/JDBC-coding-tips.html#GUID-18F42BF6-AACA-46A7-B326-F6F5F1742763) states

> Oracle JDBC drivers provide full support for, and are highly optimized for, applications that use Java multithreading. Controlled serial access to a connection, such as that provided by connection caching, is both necessary and encouraged. However, Oracle strongly discourages sharing a database connection among multiple threads. Avoid allowing multiple threads to access a connection simultaneously. If multiple threads must share a connection, use a disciplined begin-using/end-using protocol.

Reply:
A lot of water has passed under the bridge since these changes were made, and Oracle 12.1 -> 12.2 JDBC was a major change that moved to Java New IO and changed behaviour around interrupts in particular.

Ultimately, the changes made to Atlas introduce only a singular change in behaviour for Oracle connections - we will override the Hikari behaviour introduced in 3.3.0 (Gotham uses 3.2.0) to ensure java.sql.TimeoutException is not treated as an evictable event (for Oracle). Oracle explicitly throws this in response to an interrupt on the thread on which a statement is executing.

Reply:
The other changes allow configuration from clients of settings dbkvs uses - (a) namely the ability to disable checkout health tests on existing connections, and (b) the ability to specify the connection test string (or use Connection.isActive()). Ultimately, I expect we will use (a) from Gotham, which makes (b) redundant.